Add expression MLangs

bshepherdson opened the pull request:
With decently complete tests against a real DAO.

Plus some small core tweaks to fix some small bugs encountered.

bshepherdson committed:
Better error message when an "implements" target cannot be found

bshepherdson committed:
Add a foam.util.compare alongside foam.util.equals

bshepherdson committed:
ArrayDAO.select returns the original, undecorated sink

bshepherdson committed:
Overhaul MLangs to be more complete and correct - with tests!

This includes all the basic expression types from before.
- AND, OR, NOT, EQ, NEQ, LT, GT, LTE, GTE, CONTAINS(_IC), IN, and HAS.

There's no partialEval() support, and none of the sinks except COUNT have been
ported so far.

adamvy-google commented on src/core/stdlib.js:
```diff
@@ -260,6 +260,12 @@ foam.LIB({
       return a == b;
+    },
+    function compare(a, b) {
+      if ( a.compareTo ) return a.compareTo(b);
```
Did we add compareTo to the primitive types?  And if so, wont' this convert numbers into Numbers and end up being slow?

bshepherdson replied:
I'm not sure. I don't think we did. They're slow, so it's probably better to add `typeof a === "number"` checks to `equals` and `compare` to handle this.

Kevin, thoughts?

kgrgreer replied:
Yes, add a "number" check.  Also, check if Strings are slower as well, I think they might be, and then check for them if needed also.

bshepherdson replied:
I've added a special case when both values have `typeof` `'number'`.

adamvy-google commented on src/lib/dao.js:
```diff
@@ -572,6 +573,8 @@ foam.CLASS({
     function put(o) {
       this.a.push(o);
+    },
+    function eof() {
```
This shouldn't be necessary, the Sink interface provides a default empty implementation of each method.

bshepherdson replied:
Ah, yep, that's true. I added `eof()` before realizing that ti didn't implement `Sink`.

Removed.

adamvy-google commented on src/lib/mlang.js:
```diff
@@ -147,108 +259,207 @@ foam.CLASS({
+    function f(o) {
+      var s1 = this.arg1.f(o).toUpperCase();
+      var s2 = this.arg2.f(o).toUpperCase();
+      return s1 ? s1.indexOf(s2) !== -1 : false;
```
If you going to guard against s1 being undefined you need to do that before you call toUpperCase()

bshepherdson replied:
Whoops, fixed.

kgrgreer replied:
Is the 'Expr' suffix redundant?  It's already in the foam.mlang.predicate package.  The code was originally written before we had packages, which add extra semantic information already.

kgrgreer replied:
This can be made much much faster by using a regex with the ignore case option.  At least leave a TODO:  See foam/apps/quickbug/DefaultQuery.js

adamvy-google commented on src/lib/mlang.js:
```diff
@@ -147,108 +259,207 @@ foam.CLASS({
+  methods: [
+    function f(obj) {
+      var value = this.arg1.f(obj);
+      var notHas = value === undefined || value === null || value === '' ||
```
Are these the right semantics?  Or should HAS be essentially hasOwnProperty()

If I want to query for things that have a property defined rather than things that are the defaultValue.  I'm not sure what is more useful or expected.

bshepherdson replied:
We talked about this a while back. `HAS` is intended to mean "has an interesting, non-NULL value, even if that value is falsy". If there's a `defaultValue` that isn't `null` or `undefined`, then `HAS` is always true and that's fine.

Put another way, `HAS()` is roughly SQL `IS NOT NULL`.

kgrgreer replied:
Database semantics rather than JS semantics.  I think this is correct.

adamvy-google replied:
It's not really JS semantics so much as FOAM semantics.

Databases don't have default values other than NULL usually.  What about a DEFINED and HAS as separate expressions, one for when an object has a set value for a property and isn't using the default, and another for when a property is an interesting value or not, regardless of defaults.

bshepherdson replied:
I'd be happy with `DEFINED` that was roughly equivalent to `hasOwnProperty`, once we had a cause for it. Right now, I'm not sure it's worth adding.

adamvy-google commented on test/js/lib/mlang.js:
```diff
@@ -0,0 +1,496 @@
+require('../../../src/lib/dao');
+require('../../../src/lib/mlang');
+
+function buildPersonDAO() {
```
I strongly prefer setup code like this to just live in the beforeEachTest.

Helper methods in tests just obfuscate what exactly you are testing and make it harder to update tests if the helper method ever gets shared between two tests.

bshepherdson replied:
When I wrote that I expected it to go in multiple different blocks. That didn't actually happen, so I've pulled this into the main describe's `beforeEach`.

bshepherdson committed:
ArrayDAO: Decorate the same ArraySink as we return to the user

Previously when no sink was provided, we created two separate ArraySinks, put
into one and returned the other, which was always empty.

adamvy-google commented on src/lib/mlang.js:
```diff
@@ -73,72 +75,182 @@ foam.CLASS({
+  abstract: true,
+  // PORT: There should be several interfaces here eventually, eg.
+  // Functor/Projectable/whatever we call f()'s interface.
+  // implements: ['ExprProtocol'],
```
Delete this for now.  ExprProtocol was a thing added for swift, we should be able to do this better.

bshepherdson replied:
Done.

adamvy-google commented on src/lib/mlang.js:
```diff
@@ -73,72 +75,182 @@ foam.CLASS({
+      return this.toMQL();
+    },
+    /** Recursively adds all inputs of an expression to an array. */
+    function collectInputs(terms) {
```
Delete this.

bshepherdson replied:
Done.

adamvy-google commented on src/lib/mlang.js:
```diff
@@ -73,72 +75,182 @@ foam.CLASS({
+    /**
+     * Converts this expression into MQL syntax, as used in eg. Gmail searches.
+     */
+    function toMQL() {
```
Delete toMQL and toSQL, we'll add them as refinements when we need them.

bshepherdson replied:
Done. Fixed `toString()` to not use `toMQL()` as well.

adamvy-google commented on src/lib/mlang.js:
```diff
@@ -73,72 +75,182 @@ foam.CLASS({
+     * expression.
+     */
+    /*
+    function pipe(sink) {
```
I would rather delete this as well and put this as a TODO in the bottom of the file, like we did for mm.js/dao.js

bshepherdson replied:
Done.

adamvy-google commented on src/lib/mlang.js:
```diff
@@ -73,72 +75,182 @@ foam.CLASS({
+
+  methods: [
+    /**
+     * Abstract projection method. Given an object, extracts the value of a
```
This isn't always a projection method, in some cases it's a predicate function, in others it's a projection.

bshepherdson replied:
Adjusted the comment.

adamvy-google commented on src/core/stdlib.js:
```diff
@@ -289,6 +289,14 @@ foam.LIB({
       return a == b;
+    },
+    function compare(a, b) {
+      if ( typeof a === 'number' && typeof b === 'number' )
```
Add typeof a === 'string' checks.

bshepherdson replied:
Done.

adamvy-google commented on src/lib/mlang.js:
```diff
@@ -73,72 +75,182 @@ foam.CLASS({
+    },
+    /** Converts this expression to a human-readable string for debugging. */
+    function toString() {
+      return this.toMQL();
```
toString should return a EQ(AND(....)) style looking thing

bshepherdson committed:
Add toString to Property that returns the property name

bshepherdson committed:
Add toString() support to the mlangs, and test it

bshepherdson commented:
Now with `toString` support and tests for same. PTAL

kgrgreer commented on src/lib/mlang.js:
```diff
@@ -150,108 +232,259 @@ foam.CLASS({
-    function EQ(a, b) { return this._binary_("EqExpr", a, b); },
-    function GT(a, b) { return this._binary_("GtExpr", a, b); }
+    function _nary_(name, args) {
+      return this[name].create({ args: foam.array.argsToArray(args) });
```
FYI, it is faster to do:
var r = this[name].create();
r.args = foam.array.argsToArray(args);
return r;
But since the cost of creating an mLang is orders of magnitude less than using it, it doesn't matter.

adamvy-google commented:
LGTM

kgrgreer commented on src/core/stdlib.js:
```diff
@@ -0,0 +1,103 @@
+ */
+foam.LIB = function LIB(model) {
+  function defineProperty(proto, key, map) {
+    if ( ! map.value )
```
Add braces to match the coding style.

adamvy-google commented:
Can this PR be deleted now?

bshepherdson commented:
Yeah. Rebasing to sort out some conflicts made Github lose track of this PR; all the commits are merged.
